pkg/kubectl/cmd/util/clientcache.go

```diff
@@ -33,59 +34,116 @@ func NewClientCache(loader clientcmd.ClientConfig) *ClientCache {
 // ClientConfigForVersion returns the correct config for a server
-func (c *ClientCache) ClientConfigForVersion(version string) (*client.Config, error) {
-	if c.defaultConfig == nil {
-		config, err := c.loader.ClientConfig()
+func (c *ClientCache) ClientConfigForVersion(group, version string) (*client.Config, error) {
+	if _, found := c.defaultConfig[group]; !found {
+		config, err := c.loader.ClientConfig(group)
 		if err != nil {
 			return nil, err
 		}
-		c.defaultConfig = config
+		c.defaultConfig[group] = config
 		if c.matchVersion {
 			if err := client.MatchesServerVersion(c.defaultClient, config); err != nil {
 				return nil, err
 			}
 		}
 	}
-	if config, ok := c.configs[version]; ok {
+	if config, ok := c.configs[apiutil.GetGroupVersion(group, version)]; ok {
 		return config, nil
 	}
 	// TODO: have a better config copy method
-	config := *c.defaultConfig
-	negotiatedVersion, err := client.NegotiateVersion(c.defaultClient, &config, version, registered.RegisteredVersions)
+	config := *c.defaultConfig[group]
+	negotiatedVersion, err := client.NegotiateVersion(c.defaultClient, &config, group, version, registered.RegisteredVersions)
 	if err != nil {
 		return nil, err
 	}
 	config.Version = negotiatedVersion
-	client.SetKubernetesDefaults(&config)
-	c.configs[version] = &config
+	client.SetDefaultsForGroup(group, &config)
+	c.configs[apiutil.GetGroupVersion(group, version)] = &config
 
 	return &config, nil
 }
 
-// ClientForVersion initializes or reuses a client for the specified version, or returns an
+// RESTClientForVersion initializes or reuses a RESTClient for the specified version, or returns an
 // error if that is not possible
-func (c *ClientCache) ClientForVersion(version string) (*client.Client, error) {
-	if client, ok := c.clients[version]; ok {
+func (c *ClientCache) RESTClientForVersion(groupVersion string) (*client.RESTClient, error) {
+	if client, ok := c.clients[groupVersion]; ok {
 		return client, nil
 	}
-	config, err := c.ClientConfigForVersion(version)
+	config, err := c.ClientConfigForVersion(apiutil.GetGroup(groupVersion), apiutil.GetVersion(groupVersion))
 	if err != nil {
 		return nil, err
 	}
-	client, err := client.New(config)
+	client, err := client.RESTClientFor(config)
 	if err != nil {
 		return nil, err
 	}
-
+	// TODO: Version should be renamed to GroupVersion
 	c.clients[config.Version] = client
 	return client, nil
 }
+
+// ClientForVersion returns a Client with RESTClients for all groups.
+
+// TODO: Currently this function only accepts one groupVersion. It will
+// try to initialize or reuse the RESTClient of the group in the groupVersion
+// for the specified version. For RESTClients of other groups, it will initialize
+// or reuse the one of the negotiated version. An error is returned if any
+// it is not possible to return such a Client. We should make it accept a list
+// of groupVersion in the future.
+func (c *ClientCache) ClientForVersion(groupVersion string) (*client.Client, error) {
```

Comment:
I mean, client.Client *has* a version, but it's compiled into it, it doesn't change at run time, so you can't parameterize on it. You can make a different versioned RESTClient at runtime.